src/main/scala/scalaz/DistributedModule.scala

```diff
@@ -0,0 +1,58 @@
+package scalaz
+
+trait Member
+
+sealed trait Membership
+object Membership {
+  final case class Join(member: Member) extends Membership
+  final case class Leave(member: Member) extends Membership
+  final case class Unreachable(member: Member) extends Membership
+}
+
+sealed trait ValueChanged[A]
+object ValueChanged {
+  final case class Modify[A](member: Member, old: A, newV: A) extends ValueChanged[A]
+  final case class Add[A](member: Member, newV: A) extends ValueChanged[A]
+  final case class Remove[A](member: Member, value: A) extends ValueChanged[A]
+}
+
+trait DistributedModule[F[_]] {
+  type Type[A]
+  type Prism[A, B]
+  type Traversal[A, B]
+
+  sealed trait PathElem[A, B]
+  object PathElem {
+    case class Key[K, V](v: K) extends PathElem[Map[K, V], V]
+    case class MetadataId[A](v: String) extends PathElem[A, A]
+    case class Elements[A]() extends PathElem[Set[A], A]
+    case class Composed[A, B, C](l: PathElem[A, B], r: PathElem[B, C]) extends PathElem[A, C]
+  }
+
+  implicit val stringType   : Type[String]
+  implicit val longType     : Type[Long]
+  implicit val booleanType  : Type[Boolean]
+  implicit def setType[V: Type] : Type[Set[V]]
+  implicit def mapType[K: Type, V: Type]: Type[Map[K, V]]
+
+  implicit class LensSyntax[A: Type, B: Type](self: PathElem[A, B]) {
+    def >>> [C: Type](that: PathElem[B, C]): PathElem[A, C] = PathElem.Composed(self, that)
+  }
+
+  def members(callback: Membership => F[Boolean]): F[Unit]
+
+  def access[A: Type, B: Type](id: PathElem[A, B]): F[Metadata[B]]
+
+  def onChange[A: Type, B: Type](p: PathElem[A, B], callback: ValueChanged[B] => F[Unit]): F[Unit]
+
+  def key[K: Type, V: Type](k: K): PathElem[Map[K, V], V] = PathElem.Key[K, V](k)
+
+  def elements[V: Type]: PathElem[Set[V], V] = PathElem.Elements[V]()
+
+  def metadataId[A: Type](s: String): PathElem[A, A] = PathElem.MetadataId[A](s)
+
+  trait Metadata[A] {
```

Comment:
the idea with syntax is kind of weird for me. I simple don't understand how you gonna provide implementation for that and since it's implicit how you gonna make below example to compile without providing types.

`
val thisIsMe: Member = new Member {}

  val DM: DistributedModule = ???
  import DM._

  for {
    protocol <- DM.startup(thisIsMe, Set())
    map      <- protocol.access[Map[String, Set[String]]](MetadataID("regions"))
    _        <- map.set(values >>> elements, "string") // this requires couple of other changes to work but still you will end up providing type here.
  } yield ()
` 

other thing is the those are duplicated method.
def set[B: Type](where: Prism[A, B], b: b)(implicit A: Type[A]): F[Unit]
def set[B: Type](where: Lens[A, B], b: b)(implicit A: Type[A]): F[Unit]

I will keep only the one below
def set[B: Type](where: Lens[A, B], b: b)(implicit A: Type[A]): F[Unit]

Reply:
We'll they're not quite duplicated as they use different kind of optics, but that's something we will have to clarify (are lenses and prisms sufficient, do we need traversals etc.).

Reply:
sorry I should be me more explicit they are not the same for us because we know that Lens/Prism/Traversal are not the same things but they are identical for compiler after type erasure :)

I believe we need all of them - Lens, Prism, Traversal. Hoever we should think about proper design after this PR will merge. Maybe we should create ticket right away to don't forget about this.